api/src/main/java/jakarta/mail/Multipart.java

```diff
@@ -266,4 +269,27 @@ public synchronized Part getParent() {
     public synchronized void setParent(Part parent) {
         this.parent = parent;
     }
+
+    protected StreamProvider provider() {
+        if (streamProvider == null) {
+            synchronized (this) {
+                if (streamProvider == null) {
+                    if (parent != null && parent instanceof Message) {
```

Comment:
I would assume that this should loop while the parent is not null. If we find a message and it has non-null provider use it and break out.

Reply:
It should be in place now. Now the parent implements #getStreamProvider so it will deal with that.

Reply:
I realized I missed a detail in that during parsing the parent may not be attached.  I think we need check DataSource for messageaware and get session (which has an accessor for streamprovider) from messagecontext in the subclasses.

This should help parsing.

Reply:
I am checking this:
https://github.com/jakartaee/mail-api/blob/bf2bfc18c0d2b9b3f57417203c2a462a3a94a7de/api/src/main/java/jakarta/mail/internet/MimeMultipart.java#L261

It looks this is already obtained and set it as a parent. So we can just ask the parent for the `StreamProvider` instance